rotkehlchen/chain/ethereum/graph.py

```diff
@@ -77,10 +91,42 @@ def query(
 
         querystr = prefix + querystr
         log.debug(f'Querying The Graph for {querystr}')
-        try:
-            result = self.client.execute(gql(querystr), variable_values=param_values)
-        except (requests.exceptions.RequestException, Exception) as e:
-            raise RemoteError(f'Failed to query the graph for {querystr} due to {str(e)}') from e
+
+        retries_left = QUERY_RETRY_TIMES
+        while retries_left > 0:
+            try:
+                result = self.client.execute(gql(querystr), variable_values=param_values)
+            except (
+                TransportProtocolError,
+                TransportQueryError,
+                TransportServerError,
+            ) as e:
+                exc_msg = e.errors if isinstance(e, TransportQueryError) else str(e)
+                data = e.data if isinstance(e, TransportQueryError) else None
+                base_msg = f'The Graph query: {querystr} failed due to: {exc_msg}'
+
+                retries_left -= 1
+                if retries_left:
+                    sleep_seconds = RETRY_BACKOFF_FACTOR * pow(2, QUERY_RETRY_TIMES - retries_left)
+                    retry_msg = (
+                        f'Retrying query after {sleep_seconds} seconds. '
+                        f'Retries left: {retries_left}.'
+                    )
+                    log.error(f'{base_msg}. {retry_msg}', data=data)
+                    gevent.sleep(sleep_seconds)
+                else:
+                    raise RemoteError(f'{base_msg}. No retries left.') from e
+
+            except (
+                TransportAlreadyConnected,
```

Comment:
What I've understood is that it's not worth it to retry if the transport is already connected or already closed. But I could be wrong and these 2 exceptions could go up.